tonilopezmr/day4/PassphraseValidator.kt

```diff
@@ -0,0 +1,73 @@
+package day4
+
+import org.junit.Assert.*
+import org.junit.Test
+import java.io.File
+import java.util.*
+
+class PassphraseValidator {
+
+  fun validI(passphrase: String): Boolean {
+    val words = passphrase
+        .split(" ")
+    return words
+        .foldIndexed(true) { idx, acc, a ->
+          if (words.subList(idx + 1, words.size).contains(a)) {
+            false
+          } else {
+            acc
+          }
+        }
+  }
+
+  fun valid(passphrase: String): Boolean {
+    val list = LinkedList(passphrase.split(" "))
+    val element = list.poll()
+    return validR(list, element)
+  }
+
+  tailrec fun validR(passphrase: LinkedList<String>, element: String): Boolean =
+      if (passphrase.isEmpty()) true
+      else if (passphrase.contains(element)) false
+      else {
+        val element = passphrase.poll()
+        validR(passphrase, element)
+      }
```

Comment:
Why are you using `tailrec`? Did the IDE suggested it to you or you knew you had to use it?

Reply:
No! tailrec is for recursion optimization, tailrec coerce to call the method only in the last operation in `return` https://kotlinlang.org/docs/reference/functions.html and https://medium.com/@JorgeCastilloPr/tail-recursion-and-how-to-use-it-in-kotlin-97353993e17f

Reply:
Interesting, I will search if Scala has a `tailrec` reserved word too 👍